### PR TITLE
[update] Manifest Based Authorization feature update.

### PR DESCRIPTION
**Pull request by mhatrevi**
This PR  makes the following changes to the feature:

1. Adds a 'firmware id' to the image metadata entry (IME) structure.
2. Reduces max IME limit from 128 to 127 to account for size increase due to the firmware id addition.
3. SET_AUTH_MANIFEST command validates that the IME collection does not contain duplicate firmware ids.
4. SET_AUTH_MANIFEST command sorts the IME on the firmware ids.
5. Adds a new bit to the IME flags named 'SKIP AUTHORIZATION'. If set, AUTHORIZE_AND_STASH command authorizes an IME with a matching firmware id ignoring the image hash.
6. If a firmware id is not found, AUTHORIZE_AND_STASH command returns IMAGE_NOT_AUTHORIZED status code in the command output.
7. If a firmware id is found but the image hash does not match, AUTHORIZE_AND_STASH returns IMAGE_HASH_MISMATCH status code in the command output.
8. If a firmware id is found and the image hash matches or if the 'SKIP AUTHORIZATION' flag is set for the firmware id, AUTHORIZE_AND_STASH command returns IMAGE_AUTHORIZED status code in the command output.
9. AUTHORIZE_AND_STASH command uses binary search on the firmware id to lookup an IME.

### CODE REVIEW

**Comment by mlvisaya** [comment]
Should we also add the changes to the README.md doc?

**Review by nquarton — COMMENTED** [review]
Overall implementation looks good to me. I think the runtime readme will need some updates to reflect these changes.

You might also add a couple more test cases for the sorting and binary search. They look simple and correct to me. But I know corner cases can bite you on that. I'll still approve without it, just something to consider.

File: runtime/src/authorize_and_stash.rs [hunk 1]
```diff
@@ -100,4 +112,46 @@ impl AuthorizeAndStashCmd {
...
+    ///
+    #[inline(never)]
+    fn find_metadata_entry<'a>(
+        auth_manifest_image_metadata_col: &'a AuthManifestImageMetadataCollection,
+        cmd_fw_id_bytes: &[u8; 4],
+    ) -> Option<&'a AuthManifestImageMetadata> {
+        let mut left = 0;
+        let mut right = auth_manifest_image_metadata_col.entry_count as usize;
+
+        while left < right {
```
Comment (swenson) [inline runtime/src/authorize_and_stash.rs:139]: Can we use `auth_manifest_image_metadata_col.image_metadata_list.binary_search_by_key()`?

Either way, we should add unit tests for this function.

File: runtime/src/set_auth_manifest.rs [hunk 2]
```diff
@@ -401,6 +415,34 @@ impl SetAuthManifestCmd {
             soc_ifc,
         )?;
 
+        // Sort the image metadata list by firmware ID in place. Also check for duplicate firmware IDs.
+        let slice =
+            &mut metadata_mailbox.image_metadata_list[..metadata_mailbox.entry_count as usize];
```
Comment (swenson) [inline runtime/src/set_auth_manifest.rs:420]: Why can't we use `metadata_mailbox.image_metadata_list.sort_unstable_by_key()`?

If we need to check for duplicates, we can do it after sorting.

File: runtime/src/authorize_and_stash.rs (hunk 1 above)
Comment (mhatrevi) [reply runtime/src/authorize_and_stash.rs:138]: Updated.

File: runtime/src/set_auth_manifest.rs (hunk 2 above)
Comment (mhatrevi) [reply runtime/src/set_auth_manifest.rs:420]: sort_unstable_by_key is increasing the Runtime size by 2+KB.

File: runtime/README.md [hunk 3]
```diff
@@ -790,7 +799,7 @@ Command Code: `0x4154_5348` ("ATSH")
 | --------      | -------- | ---------------
 | chksum            | u32      | Checksum over other output arguments, computed by Caliptra. Little endian. |
 | fips_status      | u32      | Indicates if the command is FIPS approved or an error.                     |
-| auth_req_result | u32      | AUTHORIZE_IMAGE: 0xDEADC0DE and DENY_IMAGE_AUTHORIZATION: 0x21523F21    |
+| auth_req_result | u32      | AUTHORIZE_IMAGE (0xDEADC0DE), IMAGE_NOT_AUTHORIZED (0x21523F21) or IMAGE_HASH_MISMATCH (0x8BFB95CB) |
```
Comment (nquarton) [inline runtime/README.md:802]: For most other commands (like signature verify commands) for a negative results, we report a mailbox command failure and put the reason in the non-fatal error code reg. Just want to call that out to be sure we want this to behave differently.

On the other hand, DPE does report a result that can be a negative result with a "mailbox complete". But that's maybe a bit of an exception since it's a whole separate submodule.

**Review by nquarton — DISMISSED** [review]

File: api/src/mailbox.rs [hunk 4]
```diff
@@ -1059,8 +1059,8 @@ impl AuthAndStashFlags {
 #[derive(Debug, AsBytes, FromBytes, PartialEq, Eq)]
 pub struct AuthorizeAndStashReq {
     pub hdr: MailboxReqHeader,
-    pub metadata: [u8; 4],
-    pub measurement: [u8; 48],
+    pub metadata: [u8; 4],     // Firmware Id.
```
Comment (mlvisaya) [inline api/src/mailbox.rs:1062]: show we just rename this as fw_id? so as not to confuse with metadata in "image metadata entry"

File: runtime/src/set_auth_manifest.rs (hunk 2 above)
Comment (swenson) [reply runtime/src/set_auth_manifest.rs:420]: in that case, we should add some tests for this, and we can make it more efficient later

**Review by swenson — COMMENTED** [review]
Looks good, but definitely I would want some tests on the sorting routine.

File: runtime/src/set_auth_manifest.rs [hunk 5]
```diff
@@ -401,6 +415,34 @@ impl SetAuthManifestCmd {
...
+        for i in 0..slice.len() {
+            for j in 0..(slice.len() - 1 - i) {
+                // This check is needed to avoid out of bounds access panic.
+                if (j + 1) >= slice.len() {
+                    break;
+                }
+
+                match slice[j].fw_id.cmp(&slice[j + 1].fw_id) {
+                    core::cmp::Ordering::Greater => {
+                        slice.swap(j, j + 1);
```
Comment (swenson) [inline runtime/src/set_auth_manifest.rs:430]: If you use insertion sort instead (just changing the indices a little), then this runs something like 4x faster.

Comment (mhatrevi) [reply runtime/src/set_auth_manifest.rs]: Updated. Also added unit tests for this.

File: api/src/mailbox.rs (hunk 4 above)
Comment (mhatrevi) [reply api/src/mailbox.rs:1062]: Makes sense, updated.

File: runtime/README.md (hunk 3 above)
Comment (mhatrevi) [reply runtime/README.md:802]: Ack.

**Comment by mhatrevi** [comment]
> Looks good, but definitely I would want some tests on the sorting routine.

Added unit tests for the same.

**Review by nquarton — APPROVED** [review]

**Review by swenson — APPROVED** [review]
LGTM

File: runtime/src/authorize_and_stash.rs [hunk 6]
```diff
@@ -100,4 +112,31 @@ impl AuthorizeAndStashCmd {
...
+    ///   otherwise returns `None`.
+    ///
+    #[inline(never)]
+    fn find_metadata_entry(
+        auth_manifest_image_metadata_col: &AuthManifestImageMetadataCollection,
+        cmd_fw_id: u32,
+    ) -> Option<&AuthManifestImageMetadata> {
+        auth_manifest_image_metadata_col
+            .image_metadata_list
+            .binary_search_by(|metadata| metadata.fw_id.cmp(&cmd_fw_id))
```
Comment (jhand2) [inline runtime/src/authorize_and_stash.rs:138]: Are the entries sorted? I would expect them not to be.

File: runtime/src/set_auth_manifest.rs [hunk 7]
```diff
@@ -401,6 +415,45 @@ impl SetAuthManifestCmd {
             soc_ifc,
         )?;
 
+        // Sort the image metadata list by firmware ID in place. Also check for duplicate firmware IDs.        let slice =
+        let slice =
+            &mut metadata_mailbox.image_metadata_list[..metadata_mailbox.entry_count as usize];
+
+        Self::sort_and_check_duplicate_fwid(slice)?;
```
Comment (jhand2) [inline runtime/src/set_auth_manifest.rs:422]: This seems unnecessarily complicated. I can't imagine linear search being noticeably slow here.

File: runtime/src/authorize_and_stash.rs (hunk 6 above)
Comment (mhatrevi) [reply runtime/src/authorize_and_stash.rs:138]: They are sorted in the SET_AUTH_MANIFEST command processing.

File: runtime/src/set_auth_manifest.rs (hunk 7 above)
Comment (mhatrevi) [reply runtime/src/set_auth_manifest.rs:422]: Given we had 128 (now 127) hashes, there was some concern about the search runtimes.

**Review by jhand2 — APPROVED** [review]
